pkg/spdk/engine.go

```diff
@@ -20,233 +22,340 @@ import (
+			return err
+		}
+		if e.Port == int32(port) {
+			break
 		}
-		res.Port = int32(port)
 	}
-	if res.Port == 0 {
-		return nil, fmt.Errorf("cannot detect the port from Nvmf subsystem for engine %s", name)
+	if port == 0 || e.Port != int32(port) {
+		return fmt.Errorf("cannot find a matching listener with port %d from Nvmf subsystem for engine %s", e.Port, e.Name)
 	}
 
-	bdevRaidList, err := spdkClient.BdevRaidGet(name, 0)
-	if err != nil {
-		return nil, err
-	}
-	if len(bdevRaidList) != 1 {
-		return nil, fmt.Errorf("found multiple or zero raid bdevs in engine %v creation: %+v", name, bdevRaidList)
+	switch e.Frontend {
+	case types.FrontendSPDKTCPBlockdev:
+		initiator, err := nvme.NewInitiator(e.VolumeName, nqn, nvme.HostProc)
+		if err != nil {
+			return err
+		}
+		if err := initiator.LoadNVMeDeviceInfo(); err != nil {
+			return err
+		}
+
+		if err := initiator.LoadEndpoint(); err != nil {
+			return err
+		}
+		blockDevEndpoint := initiator.GetEndpoint()
+		if e.Endpoint != "" && e.Endpoint != blockDevEndpoint {
+			return fmt.Errorf("found mismatching between engine endpoint %s and actual block device endpoint %s for engine %s", e.Endpoint, blockDevEndpoint, e.Name)
+		}
+		e.Endpoint = blockDevEndpoint
+	case types.FrontendSPDKTCPNvmf:
+		nvmfEndpoint := GetNvmfEndpoint(nqn, e.IP, e.Port)
+		if e.Endpoint != "" && e.Endpoint != nvmfEndpoint {
+			return fmt.Errorf("found mismatching between engine endpoint %s and actual nvmf endpoint %s for engine %s", e.Endpoint, nvmfEndpoint, e.Name)
+		}
+		e.Endpoint = nvmfEndpoint
+	default:
+		return fmt.Errorf("unknown frontend type %s", e.Frontend)
 	}
-	bdevRaid := bdevRaidList[0]
-	bdevRaidInfo := bdevRaid.DriverSpecific.Raid
-	res.Uuid = bdevRaid.UUID
-	res.SpecSize = bdevRaid.NumBlocks * uint64(bdevRaid.BlockSize)
 
-	bdevNvmeList, err := spdkClient.BdevNvmeGet("", 0)
-	if err != nil {
-		return nil, err
+	bdevRaid := bdevMap[e.Name]
+	if spdktypes.GetBdevType(bdevRaid) != spdktypes.BdevTypeRaid {
+		return fmt.Errorf("cannot find a raid bdev for engine %v", e.Name)
 	}
-	bdevNvmeMap := map[string]spdktypes.BdevInfo{}
-	for _, bdevNvme := range bdevNvmeList {
-		bdevNvmeMap[bdevNvme.Name] = bdevNvme
+	bdevRaidSize := bdevRaid.NumBlocks * uint64(bdevRaid.BlockSize)
+	if e.SpecSize != bdevRaidSize {
+		return fmt.Errorf("found mismatching between engine spec size %d and actual raid bdev size %d for engine %s", e.SpecSize, bdevRaidSize, e.Name)
 	}
 
-	// TODO: Verify Mode
-	for _, baseBdev := range bdevRaidInfo.BaseBdevsList {
-		bdevNvme, exists := bdevNvmeMap[baseBdev.Name]
-		if !exists {
-			// This replica must be a local lvol
-			replicaName := spdktypes.GetLvolNameFromAlias(baseBdev.Name)
-			res.ReplicaAddressMap[replicaName] = ""
-			res.ReplicaModeMap[replicaName] = spdkrpc.ReplicaMode_RW
+	for replicaName, bdevName := range e.ReplicaBdevNameMap {
+		mode, err := e.validateAndUpdateReplicaMode(replicaName, bdevMap[bdevName])
```

Comment:
bdevName is `test-block-1/pvc-39468503-106d-4606-9f1d-2f4879d188b5-r-0aaa43fd`.
But bdeMap's key is `pvc-39468503-106d-4606-9f1d-2f4879d188b5-r-0aaa43fd`.

Reply:
For a lvol bdev, the full bdev name or the alias is `<LVS NAME>/<LVOL NAME>`. Since it's trivial to get the full name  from a replica/lvol name (the missing part is lvs name), the engine will retain `replicaBdevNameMap`

Reply:
Yes, the `bdevMap` is created from https://github.com/longhorn/longhorn-spdk-engine/pull/25#discussion_r1194833357, so `bdevMap[bdevName]` gets nothing. Should it be `replicaName`?

Reply:
OH, YES. That's the cause. Let me fix it.

Reply:
Thank you. No questions except for the issue.

Reply:
Fixed. Now `bdevMap` uses lvol bdev alias as key. 
BTW, the spdk rpc server get API will use UUID as lvol **bdev name**. For consistency & readibility, we use alias instead.